### PR TITLE
Add lesson plan for Collaborating with GitHub session Feb 22

### PR DESCRIPTION
**Pull request by sjeblee**

### CODE REVIEW

File: lessons/git/collaboration/lesson.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
+---
+layout: page
+title:
+visible: false
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: Set to `true`.

File: lessons/git/collaboration/lesson.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+tags: 
+  # languages
+  - git
+  # levels
+  - beginner
+---
+<!-- change visible to true if you want it on the site -->
+<!-- remove any tags listed above that are not relevant -->
+
+# Collaborating on GitHub
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: Put this in `title:` above and remove this line instead.

File: lessons/git/collaboration/lesson.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+ - **Authors**: Serena Jeblee
+ - **Research field**: Computational Linguistics
+ - **Lesson topic**: Collaborating on GitHub
+ - **Lesson content URL**: <link>
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: https://github.com/UofTCoders/studyGroup/tree/lessons/git/collaboration/

File: lessons/git/collaboration/lesson.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+- Some helpful git commands
+- Help! I've made a mistake! (aka how to unstage changes and undo commits)
+- Merge conflicts
+
+If we have time:
+- Stashing changes
+- Branches
+- Rebasing
+
+## Whirlwind Overview of Github
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: GitHub

File: lessons/git/collaboration/lesson.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+- What is GitHub?
+- What is version control?
+    - version control helps you collaborate on code especially when your changes might overlap
+    - master repo vs local repos
+    - master - shared version stored on github or a server
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: GitHub

File: lessons/git/collaboration/lesson.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    - local - your local copy that you can play with and make changes to, only seen by you until you push your changes
+- How to contribute code
+~~~
+git clone [repo-url]
+git pull // syncs latest version from the master repo to your local repo
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: `#` for shell comments.

File: lessons/git/collaboration/lesson.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+git push // sync your local changes to the master repo
+~~~
+- Fork and pull request if you can't push directly
+- Some rules for good commits
+    - commit should represent one change or a small set of related changes (like a bug fix or new feature)
+    - all committed code should be functional and tested
+    - make sure the commit message is descriptive
+
+## Code commit process
+- Create a new repo on github
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: GitHub

File: lessons/git/collaboration/lesson.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+## Help! I messed up and now nothing works!
+- Add junk code to hello.sh and run it
+- Undo the changes
+~~~
+    git checkout -- [file]
+~~~
+- Unstage the changes, but keep them
+~~~
+    git reset --hard HEAD and git reset HEAD~x
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: Split lines; `and` doesn't mean anything in shell.

File: lessons/git/collaboration/lesson.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+~~~
+- Hard vs soft reset
+
+## Oops, I didn't mean to commit that
+- Add the junk code back and commit it
+- Undo a local commit
+~~~
+    git reset --soft HEAD~1 \\ keeps changes, but undoes the commit
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: Slashes are backward, but use `#` instead.

**Review by QuLogic — COMMENTED** [review]
I'm not sure how others feel, but I think this could use a smidge more GitHub and a tiny bit less git review. What to do with PRs, commenting, merging, reviewing, etc.

**Comment by joelostblom** [comment]
I agree that it would be great with more information around PRs, including how to review them and go back and forth with comments and changes to the PR before merging it!

File: lessons/git/collaboration/lesson.md [hunk 10]
```diff
@@ -11,103 +11,80 @@ tags:
 <!-- change visible to true if you want it on the site -->
 <!-- remove any tags listed above that are not relevant -->
 
-# Collaborating on GitHub
-
  - **Authors**: Serena Jeblee
  - **Research field**: Computational Linguistics
  - **Lesson topic**: Collaborating on GitHub
- - **Lesson content URL**: <link>
+ - **Lesson content URL**: https://github.com/UofTCoders/studyGroup/tree/lessons/git/collaboration/
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: The `<>` are still needed to make this a link.

File: lessons/git/collaboration/lesson.md [hunk 11]
```diff
@@ -118,3 +95,31 @@ git commit
 git pull
 ~~~
 - Resolve the merge conflict, delete >>> <<< lines, add and commit
+
+## Branches
+- Branches allow you to have different commit histories in the same repo
+- Useful for feature work
+- For more info see: https://github.com/UofTCoders/studyGroup/blob/gh-pages/lessons/git/branches/lesson.md
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: Also add `<>` around this link.

File: lessons/git/collaboration/lesson.md [hunk 12]
```diff
@@ -118,3 +95,31 @@ git commit
...
+- Provide constructive suggestions for how to fix it
+
+## Updating the Pull Request
+- Try to address all comments, if you're not going to change something, say why not
+- Make a new commit and push it to the branch your pull request is on
+
+## Merging the Pull Request
+- Requires push access to the repo
+- If there are conflicts, you'll have to check out the pull request and resolve them, otherwise you can merge online
+- For more info on how to merge a pull request, see https://help.github.com/articles/merging-a-pull-request/
```
Comment (QuLogic) [inline lessons/git/collaboration/lesson.md]: `<>` here too.

**Review by QuLogic — COMMENTED** [review]
Looks good; hope we have time for a demo/hands-on of the PR stuff.

**Comment by sjeblee** [comment]
Yes, the plan is to do a hands-on exercise with pull requests.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**